File: src/tui/host.ts
// What popovers and dialogs need from the app.
import type { Editor } from "../core/editor";
import type { ExportConfig } from "../core/export";
import type { Config, GridStyle, ThemeChoice } from "../storage/config";
import type { DrawingInfo } from "../storage/drawings";
import type { ItemId, PanelId } from "./toolbar";

export interface InputDialog {
  kind: "input";
  title: string;
  label: string;
  value: string;
  cursor: number;
  error: string | null;
  /** Returns an error message to keep the dialog open, or null to close it. */
  submit(value: string): string | null | Promise<string | null>;
}

export interface ConfirmDialog {
  kind: "confirm";
  title: string;
  message: string;
  yes: string;
  confirm(): void;
}

export type Dialog = InputDialog | ConfirmDialog;

export interface Host {
  readonly editor: Editor;
  readonly config: Config;
  readonly drawingName: string;
  readonly drawings: DrawingInfo[];
  readonly currentPath: string;
  panel: PanelId | null;

  closePanel(): void;
  activate(id: ItemId, anchorX: number): void;
  openDialog(d: Dialog): void;
  closeDialog(): void;
  toast(message: string): void;
  requestRender(): void;

  // files
  openDrawing(path: string): void;
  newDrawing(): void;
  renameDrawing(): void;
  forkDrawing(): void;
  deleteDrawing(): void;
  clearDrawing(): void;
  importText(): void;

  // export
  setExport(patch: Partial<ExportConfig>): void;
  exportPreview(): string;
  copyExport(): void;
  saveExport(): void;

  // settings
  setGrid(g: GridStyle): void;
  setTheme(t: ThemeChoice): void;
  setCopyOnSelect(on: boolean): void;
  recenter(): void;
}

File: src/tui/host.rs
//! What popovers and dialogs need from the app.
//!
//! Popovers only read; every mutation happens through a [`crate::tui::painter::Action`]
//! that the app interprets after the frame. That replaces the JS closures.

use std::path::Path;

use crate::core::editor::Editor;
use crate::storage::config::Config;
use crate::storage::drawings::DrawingInfo;
use crate::tui::painter::PanelId;

pub trait Host {
    fn editor(&self) -> &Editor;
    fn config(&self) -> &Config;
    fn drawings(&self) -> &[DrawingInfo];
    fn current_path(&self) -> &Path;
    fn drawing_name(&self) -> &str;
    fn panel(&self) -> Option<PanelId>;
    fn export_preview(&self) -> String;
}

/// Which prompt an input dialog is asking for; the app owns the handler.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum InputKind {
    NewDrawing,
    RenameDrawing,
    ForkDrawing,
    ImportText,
    SaveExport,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct InputDialog {
    pub kind: InputKind,
    pub title: String,
    pub label: String,
    pub value: String,
    pub cursor: usize,
    pub error: Option<String>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ConfirmKind {
    DeleteDrawing,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ConfirmDialog {
    pub kind: ConfirmKind,
    pub title: String,
    pub message: String,
    pub yes: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Dialog {
    Input(InputDialog),
    Confirm(ConfirmDialog),
}

impl Dialog {
    pub fn title(&self) -> &str {
        match self {
            Dialog::Input(d) => &d.title,
            Dialog::Confirm(d) => &d.title,
        }
    }

    pub fn input(&self) -> Option<&InputDialog> {
        match self {
            Dialog::Input(d) => Some(d),
            Dialog::Confirm(_) => None,
        }
    }

    pub fn input_mut(&mut self) -> Option<&mut InputDialog> {
        match self {
            Dialog::Input(d) => Some(d),
            Dialog::Confirm(_) => None,
        }
    }
}
